elasticdl/doc/async_sgd_design.md

```diff
@@ -0,0 +1,118 @@
+# Design Doc: Asynchronous SGD
+
+## Motivation
+Parameter server (PS) based distributed training uses data parallelism to speed up training. 
+
+We have implemented synchronous SGD in ElasticDL. When PS accumulates `grads_to_wait` gradients from workers, PS averages these gradients and updates the model with the averaged gradients. PS also maintains `model_version` which equals to the number of model updates. Each worker has a local copy of the model. Before a minibatch training step starts, if there is a new `model_version` on PS, the worker will get the new model from PS to replace the local model. After computing the gradients with a minibatch, the worker reports the gradients to PS together with the local model version. When PS receives gradients from a worker, it only accepts the gradients with a model version same as the current PS `model_version`.
+
+
+
+This Synchronous SGD ensures model consistency in the price of wasted and blocked computation. 
+
+* Wasted computation: when a worker reports gradients with an outdated model version to PS, PS will reject these gradients. The worker have to get the current model from PS, reuse the minibatch data to train the model again.
+* Blocked computation: PS has to use a lock for model update with gradients and model read by workers to ensure model consistency.
+
+Asynchronous SGD can avoid the wasted and blocked computation mentioned above with a relaxed model consistency.
+
+* PS will accept all gradients from workers.
+* PS does not use locks and supports concurrent model reads and updates.
+
+
+## Asynchronous SGD
+Let us recall how workers train the model in synchronous SGD. Below is the pseudocode:
+
+```
+for minibatch in training_data:
+    accepted = False
+    while not accepted:
+        local_model，model_version = get_model_from_ps()
+        gradients = compute_gradient(local_model, minibatch)
+        accepted = report_gradient_to_ps(gradients, model_version)  
```

Comment:
```
accepted = report_gradient_to_ps(gradients, model_version)
```
If worker report gradient gets success, will `accpeted` be true? And then, while-loop condition is not meeted, it will finish.

Reply:
If will get the next minibatch.